// @irys/upload-solana/dist/cjs/token.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const bundles_1 = require("@irys/bundles");
const bignumber_js_1 = tslib_1.__importDefault(require("bignumber.js"));
const bs58_1 = tslib_1.__importDefault(require("bs58"));
const tweetnacl_1 = tslib_1.__importDefault(require("tweetnacl"));
const base_1 = require("@irys/upload/tokens/base");
const async_retry_1 = tslib_1.__importDefault(require("async-retry"));
const web3_js_1 = require("@solana/web3.js");
class SolanaConfig extends base_1.BaseNodeToken {
    constructor(config) {
        super(config);
        this.minConfirm = 1;
        this.finality = 'finalized';
        this.base = ['lamports', 1e9];
        this.finality = this?.opts?.finality ?? 'finalized';
    }
    async getProvider() {
        if (!this.providerInstance) {
            this.providerInstance = new web3_js_1.Connection(this.providerUrl, {
                confirmTransactionInitialTimeout: 60_000,
                commitment: this.finality,
            });
        }
        return this.providerInstance;
    }
    getKeyPair() {
        let key = this.wallet;
        if (typeof key !== 'string') {
            key = bs58_1.default.encode(Buffer.from(key));
        }
        return web3_js_1.Keypair.fromSecretKey(bs58_1.default.decode(key));
    }
    async getTx(txId) {
        const connection = await this.getProvider();
        const stx = await connection.getTransaction(txId, {
            commitment: this.finality,
            maxSupportedTransactionVersion: 0,
        });
        if (!stx)
            throw new Error('Confirmed tx not found');
        const currentSlot = await connection.getSlot(this.finality);
        if (!stx.meta)
            throw new Error(`Unable to resolve transaction ${txId}`);
        const amount = new bignumber_js_1.default(stx.meta.postBalances[1]).minus(new bignumber_js_1.default(stx.meta.preBalances[1]));
        const staticAccountKeys = stx.transaction.message.getAccountKeys().staticAccountKeys;
        const tx = {
            from: staticAccountKeys[0].toBase58(),
            to: staticAccountKeys[1].toBase58(),
            amount: amount,
            blockHeight: new bignumber_js_1.default(stx.slot),
            pending: false,
            confirmed: currentSlot - stx.slot >= 1,
        };
        return tx;
    }
    async ownerToAddress(owner) {
        return bs58_1.default.encode(owner);
    }
    async sign(data) {
        return await (await this.getSigner()).sign(data);
    }
    getSigner() {
        const keyp = this.getKeyPair();
        const keypb = bs58_1.default.encode(Buffer.concat([Buffer.from(keyp.secretKey), keyp.publicKey.toBuffer()]));
        return new bundles_1.HexSolanaSigner(keypb);
    }
    verify(pub, data, signature) {
        return bundles_1.HexSolanaSigner.verify(pub, data, signature);
    }
    async getCurrentHeight() {
        return new bignumber_js_1.default((await (await this.getProvider()).getEpochInfo()).blockHeight ?? 0);
    }
    async getFee(amount, to, multiplier) {
        const connection = await this.getProvider();
        const unsignedTx = await this._createTxUnsigned(amount, to ?? 'DHyDV2ZjN3rB6qNGXS48dP5onfbZd3fAEz6C5HJwSqRD');
        const computeBudget = new bignumber_js_1.default((await unsignedTx.getEstimatedFee(connection)) ?? 5000);
        const recentPrio = await connection
            ?.getRecentPrioritizationFees?.()
            .catch((_) => [{ prioritizationFee: 0 }]);
        const prioAvg = recentPrio
            .reduce((n, p) => n.plus(p.prioritizationFee), new bignumber_js_1.default(0))
            .dividedToIntegerBy(recentPrio.length ?? 1);
        return {
            computeBudget,
            computeUnitPrice: prioAvg
                .multipliedBy(multiplier ?? 1)
                .integerValue(bignumber_js_1.default.ROUND_CEIL),
        };
    }
    async sendTx(data) {
        const connection = await this.getProvider();
        try {
            return await (0, web3_js_1.sendAndConfirmTransaction)(connection, data, [this.getKeyPair()], { commitment: this.finality });
        }
        catch (e) {
            if (e.message.includes('30.')) {
                const txId = e.message.match(/[A-Za-z0-9]{87,88}/g);
                if (!txId)
                    throw e;
                try {
                    const conf = await connection.confirmTransaction({
                        signature: txId[0],
                        blockhash: data.recentBlockhash,
                        lastValidBlockHeight: data.lastValidBlockHeight,
                    }, this.finality);
                    if (conf)
                        return undefined;
                    throw {
                        message: e.message,
                        txId: txId[0],
                    };
                }
                catch (e) {
                    throw {
                        message: e.message,
                        txId: txId[0],
                    };
                }
            }
            throw e;
        }
    }
    async _createTxUnsigned(amount, to, fee) {
        const keys = this.getKeyPair();
        const blockHashInfo = await (0, async_retry_1.default)(async (bail) => {
            try {
                return await (await this.getProvider()).getLatestBlockhash(this.finality);
            }
            catch (e) {
                if (e.message?.includes('blockhash'))
                    throw e;
                else
                    bail(e);
                throw new Error('Unreachable');
            }
        }, { retries: 3, minTimeout: 1000 });
        const transaction = new web3_js_1.Transaction({
            ...blockHashInfo,
            feePayer: keys.publicKey,
        });
        transaction.add(web3_js_1.SystemProgram.transfer({
            fromPubkey: keys.publicKey,
            toPubkey: new web3_js_1.PublicKey(to),
            lamports: +new bignumber_js_1.default(amount).toNumber(),
        }));
        if (!this?.config?.opts?.disablePriorityFees && fee) {
            transaction.add(web3_js_1.ComputeBudgetProgram.setComputeUnitPrice({
                microLamports: fee.computeUnitPrice.toNumber(),
            }));
            transaction.add(web3_js_1.ComputeBudgetProgram.setComputeUnitLimit({
                units: fee.computeBudget.toNumber(),
            }));
        }
        return transaction;
    }
    async createTx(amount, to, fee) {
        const keys = this.getKeyPair();
        const unsignedTx = await this._createTxUnsigned(amount, to, fee);
        const transactionBuffer = unsignedTx.serializeMessage();
        const signature = tweetnacl_1.default.sign.detached(transactionBuffer, keys.secretKey);
        unsignedTx.addSignature(keys.publicKey, Buffer.from(signature));
        return { tx: unsignedTx, txId: undefined };
    }
    getPublicKey() {
        const key = this.getKeyPair();
        return key.publicKey.toBuffer();
    }
}
exports.default = SolanaConfig;
